Assets

paulsonnentag opened the pull request:
WIP

geoffreylitt commented:
Schema for assets storage looks good to me! Experience feels good to if you can iron out the issues with service worker reliability on first load I think the experience.

A couple comments on the URL design:

1) It feels odd to me having the doc type in the path after the ID. It doesn't feel like a "subpath" conceptually, more like a query param for interpreting the doc. We also need to have space to add a branch parameter for patchwork URLs. With those two things in mind, what about `http://localhost:5173/my-cool-page-2YamV2zkS2whjLJ3hTw73rhXUGXv?docType=essay` as the URL? It seems like that'd also make the trailrunner-style routing possible too?

2) I think I support the move to remove `automerge:` but not sure. I like having the URL be shorter. And it seems nice to me to not have the word `automerge` in every URL. But I think in the past @pvh has had some arguments for keeping that there, IIRC to always make it clear when we're dealing with an automerge url in any context.

3) Notably if we ship this to only TEE-main and not patchwork, it will break the ability to just add `patchwork.` to any `tee.` URL and have it just work. Maybe we shouldn't block merging this until we resolve that, but I am curious how much work it'll be to port over. I think handling the branch parameter is the main thing, I also did some refactoring of URLs and selection stuff in Patchwork which isn't in main yet.

Also: do you know the maximum image size that works before things break down? Might be worth testing that and capping the upload or at least warning.

geoffreylitt commented:
I guess another idea on the URLs might be to move doc type into the documents themselves and then remove it from the URL. But that’s probably a future step rather than something for this change.

paulsonnentag commented:
One idea as a workaround to support quickly switching between tee and patchwork: What if you can add &patchwork to the url to trigger a redirect to the document in patchwork?

paulsonnentag commented:
- I've adopted the URL schema @geoffreylitt suggested. The only drawback is that now we can't use the native file resolution of the browser because the URLs are resolved relative to to localhost now. I think this is fine, though. Thinking about it more, I think it's more incidental that this worked before (because we are only rendering a single document)
- I've fixed the known issues with the service worker and image loading
- I've changed the format to insert image tags instead of markdown references so we can specify the dimensions of the image to avoid layout shifting when the image is being loaded
- There is still some flickering with the images when they are reloaded. This needs some more tweaking. I also want to add a loading animation (maybe a light pulsating background)

geoffreylitt commented on service-worker.js:
```diff
@@ -0,0 +1,188 @@
+      })()
+    );
+  }
+  // disable caching other files for now
```
should we just remove this code for now?

paulsonnentag replied:
Maybe we could add a condition to only enable it in production. If it leads to issues, we could still remove it. Otherwise we still don't have real offline support

geoffreylitt commented on src/DocExplorer/components/DocExplorer.tsx:
```diff
@@ -50,21 +55,21 @@ export const DocExplorer: React.FC = () => {
-  const selectedDocLink = rootFolderDoc?.docs.find(
-    (doc) => doc.url === selectedDocUrl
-  );
+  const [selectedDoc] = useDocument(selectedDocLink?.url);
```
curious what the purpose of pulling this out was!

paulsonnentag replied:
I like to keep hooks simple. useSelectedDocLink doesn't use the doc and it's straight forward to call useDocument in the actual component.

geoffreylitt commented on src/DocExplorer/components/DocExplorer.tsx:
```diff
@@ -273,89 +306,112 @@ const parseCurrentUrlHash = (): UrlHashParams => {
+}): [DocLink, (docLink: DocLink) => void] => {
+  const currentUrl = useCurrentUrl();
+
+  const setSelectedDocLink = (docLink: DocLink) => {
```
Just to confirm: does all routing still flow thru setting the URL and then responding to that? (It looks like it does but I'm skimming quickly and can't tell)

I think it's important to preserve that property for a specific reason: it lowers the risk of the URL and displayed doc deviating -- which would be bad for security if I copy-pasted the URL and sent it to someone and it was the wrong doc

paulsonnentag replied:
Yes, the useCurrentUrl intercepts all navigation events and updates the url state

geoffreylitt commented on src/DocExplorer/navigation.ts:
```diff
@@ -0,0 +1,46 @@
+  // Initialize the state with the current location
+  const [url, setUrl] = useState<URL>(() => new URL(location.href));
+
+  console.log();
```
plz remove

geoffreylitt commented on src/vendor/automerge-wasm/index.d.ts:
```diff
@@ -0,0 +1,313 @@
+export type Actor = string;
```
Is this all just from upgrading automerge-wasm?

paulsonnentag replied:
I've just copied that over from trail-runner

geoffreylitt commented on src/tee/codemirrorPlugins/dragAndDropImages.ts:
```diff
@@ -0,0 +1,74 @@
+    const files = event.dataTransfer.files;
+    if (files.length > 0) {
+      const file = files[0];
+      if (file.type.startsWith("image/")) {
```
is this a reliable check for images that can be displayed on the web? I feel like we should limit the formats more specifically to a few known good ones.

geoffreylitt commented on src/tee/codemirrorPlugins/dragAndDropImages.ts:
```diff
@@ -0,0 +1,74 @@
+          }
+        });
+      } else {
+        alert("Only image files are allowed.");
```
Can we specify the formats per above

geoffreylitt reviewed (commented):
Great work -- I tested in Chrome and seems to work great when it works! Really excited for this experience, it will be a huge deal for writing internal docs.

Unfortunately still seeing sync issues. Eg: when I make a doc with a few images in it, and then open it in two browsers, often I get desyncs... curious if you can repro?

Also major browser compat regressions: Safari 16.1 and Firefox 122.0 both don't work at all, I just get a white screen. (They work with the current main branch TEE.) I know you're aware of this, can we fix before merging? Losing those browsers is a pretty big blow and I'd like to avoid it if at all possible, esp because we're unlikely to revisit that unless we hold the line.

Left some minor specific comments as well
